File: python/tests/fib.py
"""The system under test: a tiny Fibonacci library.

Every omni port ships this same library, with the same behaviour and the
same error messages, and tests it with the same spec file (spec/fib.json).
"""

from voxgig_omni.util import isnum, stringify


def fibindex(val):
    """Validate a Fibonacci index. The spec matches on these messages."""
    if not isnum(val):
        raise ValueError('fib: not a number: ' + stringify(val))

    if val % 1 != 0:
        raise ValueError('fib: non-integer index: ' + stringify(val))

    if val < 0:
        raise ValueError('fib: negative index: ' + stringify(val))

    return int(val)


def fib(n):
    """The nth Fibonacci number: fib(0)=0, fib(1)=1."""
    index = fibindex(n)

    if index == 0:
        return 0

    prev, cur = 0, 1

    for _i in range(1, index):
        prev, cur = cur, prev + cur

    return cur


def fibseq(n):
    """The first n Fibonacci numbers."""
    count = fibindex(n)
    return [fib(i) for i in range(count)]


def fibrange(start, end):
    """The Fibonacci numbers from index start to index end, inclusive."""
    from_index = fibindex(start)
    to_index = fibindex(end)
    return [fib(i) for i in range(from_index, to_index + 1)]


def fibinfo(n):
    """A description of the nth Fibonacci number.

    `prev` is None at index 0, which exercises the runner's null handling.
    """
    index = fibindex(n)
    val = fib(index)

    return {
        'n': index,
        'val': val,
        'even': val % 2 == 0,
        'prev': None if index == 0 else fib(index - 1),
        'label': 'fib(' + stringify(index) + ')=' + stringify(val),
    }

File: clojure/test/voxgig/omni/test/fib.clj
;; The system under test: a tiny Fibonacci library.
;;
;; Every omni port ships this same library, with the same behaviour and the
;; same error messages, and tests it with the same spec file
;; (spec/fib.json).

(ns voxgig.omni.test.fib
  (:require [voxgig.omni.util :as u]))

;; Validate a Fibonacci index. The spec matches on these messages. The
;; library raises a plain exception, never the runner's own error type.
(defn fibindex [val]
  (when-not (u/isnum val)
    (throw (RuntimeException. (str "fib: not a number: " (u/stringify val)))))

  (let [num (double val)]
    (when-not (== num (Math/rint num))
      (throw (RuntimeException. (str "fib: non-integer index: " (u/stringify val)))))

    (when (> 0 num)
      (throw (RuntimeException. (str "fib: negative index: " (u/stringify val)))))

    (long num)))

;; The nth Fibonacci number: fib(0)=0, fib(1)=1.
(defn fibnum [index]
  (if (zero? index)
    0
    (loop [step 1 prev 0 cur 1]
      (if (>= step index)
        cur
        (recur (inc step) cur (+ prev cur))))))

(defn fib [val] (double (fibnum (fibindex val))))

;; The first n Fibonacci numbers.
(defn fibseq [val]
  (vec (map (fn [index] (double (fibnum index))) (range (fibindex val)))))

;; The Fibonacci numbers from index start to index end, inclusive.
(defn fibrange [start end]
  (let [from (fibindex start)
        to (fibindex end)]
    (vec (map (fn [index] (double (fibnum index))) (range from (inc to))))))

;; A description of the nth Fibonacci number. `prev` is nil at index 0,
;; which exercises the runner's null handling.
(defn fibinfo [val]
  (let [index (fibindex val)
        num (fibnum index)]
    (array-map
     "n" (double index)
     "val" (double num)
     "even" (zero? (mod num 2))
     "prev" (if (zero? index) nil (double (fibnum (dec index))))
     "label" (str "fib(" (u/numstr index) ")=" (u/numstr num)))))
